File: 1590.使数组和能被-p-整除.go
/*
 * @lc app=leetcode.cn id=1590 lang=golang
 *
 * [1590] 使数组和能被 P 整除
 */

// @lc code=start
func minSubarray(nums []int, p int) int {
	n := len(nums)
	pre := make([]int, n+1)
	for i := 0; i < n; i++ {
		pre[i+1] = (pre[i] + nums[i])
	}
	x := pre[n] % p
	if x == 0 {
		return 0
	}
	ans := n
	hashmap := make(map[int]int)
	for i := 0; i < n; i++ {
		y := pre[i+1] % p
		hashmap[y] = i + 1
		z := (y - x + p) % p         // 为什么要加p，因为可能出现负数
		if j, ok := hashmap[z]; ok { // go的语法糖，如果j存在，ok为true
			ans = min(ans, i+1-j)
		}
	}
	if ans < n {
		return ans
	}
	return -1
}
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// @lc code=end

File: 1590.使数组和能被-p-整除.py
#
# @lc app=leetcode.cn id=1590 lang=python3
#
# [1590] 使数组和能被 P 整除
#
from typing import List
from itertools import accumulate
# @lc code=start
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        n = len(nums)
        pre = list(accumulate(nums, initial=0))
        x = pre[n] % p
        if x == 0:
            return 0
        hashmap = {0: 0}
        ans = n
        for i in range(n):
            y = pre[i + 1] % p
            hashmap[y] = i + 1
            z = (y - x) % p
            if z in hashmap:
                ans = min(ans, i + 1 - hashmap[z])
        return ans if ans < n else -1



    

# @lc code=end

